Refactor: Relationships UI improvements

amosmachora opened the pull request:
## Requirements

- [x] This PR has a title that briefly describes the work done, including the ticket number if there is a ticket.
- [x] My work conforms to the [**OpenMRS 3.0 Styleguide**](https://om.rs/styleguide).
- [x] I checked for feature overlap with [**existing widgets**](https://om.rs/directory).


## Summary
In thi PR I have updated the UI for the relationships page on the patient chart. Previously when adding a new relationship the user was being redirect to a new page. In this PR I have replaced that functionality with a workspace which allows the user to remain in a single page improving the ux drastically.

Also I made some small changes on the labels to more accurately describing what action the button will be persorming e.g `Add PNS Contact`.

There is a small UI bug when searching for a patient that makes the results appear terrible. I will address that in a different PR. All feedback will be highly appreciated.

## Screenshots
### Screenshot A
![image](https://github.com/user-attachments/assets/45bf83d6-a3d8-4254-8096-a8739a6c0903)

### Screenshot B
![image](https://github.com/user-attachments/assets/345af94b-3586-4cdd-ab4e-72bca960c4f1)

### Other Relationships Tab Screenshot
![image](https://github.com/user-attachments/assets/817a7507-23f5-4393-8da9-88e6648695f9)

### Video
https://github.com/user-attachments/assets/469e0740-9049-4ceb-81c8-a109297be23b

<!--
Optional.
If possible, please insert any screenshots/videos of your changes here.
Don't forget to remove the *None.* above if you do fill this section.
-->


## Related Issue

*None.*
<!--
Required if applicable.
If present, please link any related issue here, e.g. "https://issues.openmrs.org/browse/123").
Don't forget to remove the *None.* above if you do fill this section.
-->


## Other

*None.*
<!--
Optional.
Anything else that isn't covered by one of the sections above.
Don't forget to remove the *None.* above if you do fill this section.
-->

makombe commented on packages/esm-patient-clinical-view-app/src/family-partner-history/family-relationship.workspace.tsx:
```diff
@@ -0,0 +1,204 @@
+
+  const onSubmit: SubmitHandler<FormData> = async (data) => {
+    const payload = {
+      personA: personAUUID,
```
```suggestion
      personA: rootPersonUuid,
```

makombe commented on packages/esm-patient-clinical-view-app/src/family-partner-history/family-relationship.workspace.tsx:
```diff
@@ -0,0 +1,204 @@
+  const onSubmit: SubmitHandler<FormData> = async (data) => {
+    const payload = {
+      personA: personAUUID,
+      personB: patientUuid,
```
```suggestion
      personB: relatedPersonUuid,
```

makombe commented on packages/esm-patient-clinical-view-app/src/family-partner-history/family-relationship.workspace.tsx:
```diff
@@ -0,0 +1,204 @@
+      mutate(relationshipsUrl);
+      showSnackbar({
+        kind: 'success',
+        title: t('saveRlship', 'Save Relationship'),
```
```suggestion
        title: t('saveRelationship', 'Save Relationship'),
```

makombe commented on packages/esm-patient-clinical-view-app/src/family-partner-history/family-relationship.workspace.tsx:
```diff
@@ -0,0 +1,204 @@
+    } catch (err) {
+      showSnackbar({
+        kind: 'error',
+        title: t('RlshipError', 'Relationship Error'),
```
```suggestion
        title: t('relationshpError', 'Relationship Error'),
```

makombe commented on packages/esm-patient-clinical-view-app/src/family-partner-history/family-relationship.workspace.tsx:
```diff
@@ -0,0 +1,204 @@
+            name="notes"
+            control={control}
+            render={({ field }) => (
+              <TextArea labelText="Any additional notes" rows={4} id="case-manager-notes" {...field} />
```
```suggestion
              <TextArea labelText={t('addiotionalNotes','Any additional notes')} rows={4} id="relationship-notes" {...field} />
```

amosmachora replied:
Thanks. I was reusing some of the code from the case manager component. That snuck through. Thanks

makombe commented on packages/esm-patient-clinical-view-app/src/family-partner-history/family-relationship.workspace.tsx:
```diff
@@ -0,0 +1,204 @@
+
+const schema = z.object({
+  relationship: z.string({ required_error: 'Relationship is required' }),
+  startDate: z.date({ required_error: 'Start Date is required' }),
```
```suggestion
  startDate: z.date({ required_error: 'Start date is required' }),
```

amosmachora replied:
Thanks did not notice that.

makombe commented on packages/esm-patient-clinical-view-app/src/family-partner-history/family-relationship.workspace.tsx:
```diff
@@ -0,0 +1,204 @@
+
+type FamilyRelationshipFormProps = {
+  closeWorkspace: () => void;
+  personAUUID: string;
```
```suggestion
  rootPersonUuid: string;
```

makombe commented on packages/esm-patient-clinical-view-app/src/family-partner-history/family-relationship.workspace.tsx:
```diff
@@ -0,0 +1,204 @@
+});
+type FormData = z.infer<typeof schema>;
+
+type FamilyRelationshipFormProps = {
```
```suggestion
type RelationshipFormProps = {
```

amosmachora replied:
I always prefer to have a similar name for the type and the component itself but thanks well received.

makombe commented on packages/esm-patient-clinical-view-app/src/family-partner-history/family-relationship.workspace.tsx:
```diff
@@ -0,0 +1,204 @@
+};
+
+export const familyRelationshipTypes = [
+  'Sibling/Sibling',
+  'Parent/Child',
+  'Aunt/Uncle/Niece/Nephew',
+  'Guardian/Dependant',
+  'Spouse/Spouse',
+  'Partner/Partner',
+  'Co-wife/Co-wife',
+  'SNS/SNS',
+  'Injectable-drug-user/Injectable-druguser',
+];
```
Can we extract this into a config and group them into family and other and use uuids

amosmachora replied:
I am not really sure how I can do that but I will text you tomorrow you help me figure it out

makombe commented on packages/esm-patient-clinical-view-app/src/family-partner-history/family-relationship.workspace.tsx:
```diff
@@ -0,0 +1,204 @@
+        id: relationship.uuid,
+        text: relationship.display,
+      }))
+      .filter((r) => familyRelationshipTypes.includes(r.text)) || [];
```
With the config above you will now not need to filter here.

amosmachora replied:
Probably. You will have to show me how to do that.

makombe commented on packages/esm-patient-clinical-view-app/src/family-partner-history/family-relationship.workspace.tsx:
```diff
@@ -0,0 +1,204 @@
+    }
+  };
+
+  const selectPatient = (patientUuid: string) => {
+    setPatientUuid(patientUuid);
```
```suggestion
  const selectPatient = (relatedPersonUuid: string) => {
    setPatientUuid(relatedPersonUuid);
```

makombe replied:
You can also name the setters appropriately.

amosmachora replied:
on it.

makombe commented on packages/esm-patient-clinical-view-app/src/family-partner-history/relationships.resource.tsx:
```diff
@@ -74,6 +76,14 @@ function mapObservations(obsData) {
 
+export const useAllRelationshipTypes = () => {
+  const customRepresentation = 'custom:(uuid,display)';
+  const url = `/ws/rest/v1/relationshiptype?v=${customRepresentation}`;
```
```suggestion
  const url = `${restBaseUrl}/relationshiptype?v=${customRepresentation}`;
```

makombe commented on packages/esm-patient-clinical-view-app/src/family-partner-history/family-relationship.scss:
```diff
@@ -0,0 +1,104 @@
+@use '@carbon/styles/scss/spacing';
+@use '@carbon/styles/scss/type';
+@use '@carbon/layout';
+@import '~@openmrs/esm-styleguide/src/vars';
```
```suggestion

```

makombe commented on packages/esm-patient-clinical-view-app/src/family-partner-history/family-relationship.scss:
```diff
@@ -0,0 +1,104 @@
+  height: 100%;
+  padding-bottom: 20px;
+}
+
+.buttonSet {
+  margin-top: auto;
+  display: flex;
+}
+
+.button {
+  height: spacing.$spacing-10;
+  display: flex;
+  align-content: flex-start;
+  align-items: center;
+  max-width: 50%;
+  min-width: 50%;
+  width: 50%;
+}
+
+.grid {
```
Try to localize the naming to avoid style overrides.

amosmachora replied:
okay got it

makombe reviewed (commented):
Thanks @amosmachora LGTM. There are some suggestions that you can address

amosmachora commented:
@makombe thanks for the feedback. I will work on them and update the PR accordingly.

amosmachora commented:
@ojwanganto I have added the other relationships tab. I have also added a workspace to facilitate adding the same. @makombe there is a small issue that I have. I want to filter out the `Relationship Types` during the API call. Currently I fetch everything the filter out client side. It works but it is not optimal.

ojwanganto commented on packages/esm-patient-clinical-view-app/src/config-schema.ts:
```diff
@@ -120,39 +120,39 @@ export const configSchema = {
       {
         uuid: '8d91a01c-c2cc-11de-8d13-0010c6dffd0f',
-        displayAIsToB: 'Sibling',
+        display: 'Sibling/Sibling',
```
+1. The config looks good

ojwanganto commented:
Thanks, @amosmachora. It's looking good. Tagging @makombe and @donaldkibet for in-depth reviews.

makombe commented on packages/esm-patient-clinical-view-app/src/family-partner-history/family-history.component.tsx:
```diff
@@ -95,15 +106,26 @@ const FamilyHistory: React.FC<FamilyHistoryProps> = ({ patientUuid }) => {
+    return (
+      <DataTableSkeleton
+        headers={headers}
+        aria-label="patient bills table"
```
Is this miss placed label?

makombe reviewed (dismissed):
LGTM. Thanks @amosmachora

amosmachora commented:
@makombe i fixed the misplaced aria label and that by proxy markd your approval as stale. Mind approving again ?

ojwanganto commented:
Thanks, @amosmachora. 
CC @makombe @donaldkibet

makombe commented:
@amosmachora Please resolve the conflicts then we can merge it in. I think it LGMT
